[18.0][MIG] product_attribute_custom_variant

rousseldenis opened the pull request:
Based on :

- https://github.com/OCA/product-attribute/pull/1839

See related module : https://github.com/OCA/sale-workflow/pull/3631

SirAionTech committed:
[ADD] product_attribute_custom_value_variant

rousseldenis committed:
[IMP] product_attribute_custom_value_variant: pre-commit autofixes

rousseldenis committed:
[MIG] product_attribute_custom_value_variant: Migration to 18.0

rousseldenis committed:
[IMP] product_attribute_custom_value_variant: Add test for constraint

rousseldenis commented:
/ocabot migration product_attribute_custom_variant

rousseldenis committed:
[IMP] product_attribute_custom_value_variant: Improve readability / split code in functions

rousseldenis commented:
@SirAionTech This can be your interest.

I've added two improvements you can maybe backport to your PR.

rousseldenis commented:
@monen17

rousseldenis commented:
I think the sale part can be splitted.

monen17 commented on product_attribute_custom_value_variant/models/sale_order_line.py:
```diff
@@ -0,0 +1,95 @@
+                in new
+            )
+            if custom_combination:
+                self._create_custom_attribute_combination(
```
```suggestion
                line._create_custom_attribute_combination(
```

monen17 commented on product_attribute_custom_value_variant/models/sale_order_line.py:
```diff
@@ -0,0 +1,95 @@
+        # Create new attribute values for each "Create custom variant" custom value
+        values_to_unlink = self.env["product.attribute.custom.value"].browse()
+        new_attribute_values = self.env["product.attribute.value"].browse()
+        # Create new attribute values for each "Create custom variant" custom value
```
```suggestion
        # Create new attribute values for each "Create custom variant" custom value
        values_to_unlink = self.env["product.attribute.custom.value"].browse()
        new_attribute_values = self.env["product.attribute.value"].browse()
```
Duplicated comment

monen17 commented on product_attribute_custom_value_variant/models/sale_order_line.py:
```diff
@@ -0,0 +1,95 @@
+            new_attribute_values, custom_values_to_unlink = (
+                self._get_new_and_unlink_custom_attribute()
+            )
+            custom_values_to_unlink |= custom_values_to_unlink
```
```suggestion
        custom_values_to_unlink = self.env["product.attribute.custom.value"].browse()
        for line in self:
            new_attribute_values, line_custom_values_to_unlink = (
                self._get_new_and_unlink_custom_attribute()
            )
            custom_values_to_unlink |= line_custom_values_to_unlink
```
I'm afraid that without a renaming (suggested above), the new `custom_values_to_unlink` would shadow the existing _out-of-the-loop_ variable creating an unwanted behavior.

monen17 commented on product_attribute_custom_value_variant/README.rst:
```diff
@@ -0,0 +1,110 @@
+   - 10
+   - Custom
+
+3. For the "Custom" value, enable custom and Create custom variant
```
```suggestion
3. For the "Custom" value, enable "Free text" and "Create custom variant"
```
The field has been renamed in https://github.com/odoo/odoo/commit/50613922fbe67a3bc7c8ccec5799ac49bed0c23b#diff-8acbc08970352949f307ee98ef3f4004f46e0f588ac1b6e82b3b40eaf91d3ebeL39-R39

monen17 commented on product_attribute_custom_value_variant/models/sale_order_line.py:
```diff
@@ -1,60 +1,87 @@
 
+    def _get_new_and_unlink_custom_attribute(
+        self,
+    ) -> tuple[ProductAttributeValue, ProductAttributeCustomValue]:
```
This is interesting, are we using types now in Odoo? Maybe I'm not updated with the latest trends, do you have any reference?

rousseldenis replied:
Typing is great to auto document the code (it's Odoo here, but it's valid for other projects) as it's standard python and does not harm the code.

monen17 reviewed (changes requested):
Thanks for the PR!
I have tried this module in a local instance and it works great!

I also have reviewed the code and noticed a few thinkgs, see the other comments.

Please edit the title of commit https://github.com/OCA/product-attribute/pull/1934/commits/38ddd110479722d4c144044740e20368fc40b7dd because it's too long according to the guidelines:
> please check if the commit message is cut with ellipsis

(from https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#71commit-message)

and gets truncated: 
![image](https://github.com/user-attachments/assets/757b5ab5-15e3-48ca-a576-ed9e78c50f4f)
Since the module name consumes a lot of space already, you can either shorten it or the commit description with something like:
> [IMP] product_attribute_custom_value_variant: Readability
>
> Split code in functions

One more thought: it might also be a **[REF]** because the module's behavior is not changing.

About the backporting, I prefer to do such things when PRs are merged for two reasons:
- avoid the overhead of keeping multiple PRs synced
- one more reason to merge PRs :smirk:

---
> I think the sale part can be splitted.

I agree that ideally this module would only depend on `product` and we'd have a _bridge_ module to make it work with `sale`, but as far as I know custom attribute values have very little (maybe none?) use without sales so the _bridge_ module would be installed very often making the decoupling useless.
This said, if someone wants to do it it's ok by me.

rousseldenis commented:
> About the backporting, I prefer to do such things when PRs are merged for two reasons:
> 
> * avoid the overhead of keeping multiple PRs synced
> * one more reason to merge PRs 😏

Yes, I approved your 16.0's PR. If you can attract people there, we can merge :-)

rousseldenis commented:
> I agree that ideally this module would only depend on `product` and we'd have a _bridge_ module to make it work with `sale`, but as far as I know custom attribute values have very little (maybe none?) use without sales so the _bridge_ module would be installed very often making the decoupling useless.
> This said, if someone wants to do it it's ok by me.

I will publish the split here soon. The idea was also to simplify the code to be called on sale side in order to replace the variant by the new one.

The idea was to be able to get new variant from custom attribute in any other flow.

rousseldenis commented:
> (from https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#71commit-message)
> 
> and gets truncated:

I agree with short commit message BUT I'm not with non meaning ones. And that's not because github interface truncates it that they should go to something like `[IMP] p_a_c_v: (...)` :sweat_smile:

rousseldenis committed:
[REF] product_attribute_custom_value_variant: Split the logic

Move the logic out of sale order model in order to simplify the calls
on that one.

Split also the tests with setup in a common class.

rousseldenis committed:
[IMP] product_attribute_custom_value_variant: Improve USAGE section
